src/lib.rs

```diff
@@ -364,49 +366,137 @@ where
+        match value {
+            TextOrBinary::Text(txt) => ws::Message::Text(txt),
+            TextOrBinary::Binary(bin) => ws::Message::Binary(bin),
+        }
+    }
+}
+
 /// Trait for encoding and decoding WebSocket messages.
 ///
 /// This allows you to customize how messages are encoded when sent over the
 /// wire.
 pub trait Codec {
-    /// The errors that can happen when using this codec.
-    type Error;
+    /// The errors that can happen when encoding using this codec.
+    type EncodeError;
+    /// The errors that can happen when decoding using this codec.
+    type DecodeError;
 
     /// Encode a message.
-    fn encode<S>(msg: S) -> Result<Vec<u8>, Self::Error>
+    fn encode<S>(msg: S) -> Result<TextOrBinary, Self::EncodeError>
     where
         S: Serialize;
 
     /// Decode a message.
-    fn decode<R>(buf: Vec<u8>) -> Result<R, Self::Error>
+    fn decode<R>(msg: TextOrBinary) -> Result<R, Self::DecodeError>
     where
         R: DeserializeOwned;
 }
 
-/// A [`Codec`] that serializes messages as JSON using `serde_json`.
+/// A [`Codec`] that serializes messages as JSON using `serde_json` and transmits it as text.
+/// Note that receiving messages works as both binary or text
 #[cfg(feature = "json")]
 #[cfg_attr(docsrs, doc(cfg(feature = "json")))]
 #[derive(Debug)]
 #[non_exhaustive]
-pub struct JsonCodec;
+pub struct TextJsonCodec;
 
 #[cfg(feature = "json")]
 #[cfg_attr(docsrs, doc(cfg(feature = "json")))]
-impl Codec for JsonCodec {
-    type Error = serde_json::Error;
+impl Codec for TextJsonCodec {
+    type EncodeError = serde_json::Error;
+    type DecodeError = serde_json::Error;
 
-    fn encode<S>(msg: S) -> Result<Vec<u8>, Self::Error>
+    fn encode<S>(msg: S) -> Result<TextOrBinary, Self::EncodeError>
     where
         S: Serialize,
     {
-        serde_json::to_vec(&msg)
+        serde_json::to_string(&msg).map(TextOrBinary::Text)
     }
 
-    fn decode<R>(buf: Vec<u8>) -> Result<R, Self::Error>
+    fn decode<R>(msg: TextOrBinary) -> Result<R, Self::DecodeError>
     where
         R: DeserializeOwned,
     {
-        serde_json::from_slice(&buf)
+        match msg {
+            TextOrBinary::Text(txt) => serde_json::from_str(&txt),
+            TextOrBinary::Binary(bin) => serde_json::from_slice(&bin),
+        }
+    }
+}
+
+/// A [`Codec`] that serializes messages as JSON using `serde_json` and transmits it as binary.
+/// Note that receiving messages works as both binary or text
+#[cfg(feature = "json")]
+#[cfg_attr(docsrs, doc(cfg(feature = "json")))]
+#[derive(Debug)]
+#[non_exhaustive]
+pub struct BinaryJsonCodec;
+
+#[cfg(feature = "json")]
+#[cfg_attr(docsrs, doc(cfg(feature = "json")))]
+impl Codec for BinaryJsonCodec {
+    type EncodeError = serde_json::Error;
+    type DecodeError = serde_json::Error;
+
+    fn encode<S>(msg: S) -> Result<TextOrBinary, Self::EncodeError>
+    where
+        S: Serialize,
+    {
+        serde_json::to_vec(&msg).map(TextOrBinary::Binary)
```

Comment:
Shouldn't that be `::Text`?

Reply:
Why would it be? This is in the impl for BinaryJsonCodec, not TextJsonCodec (where it is indeed ::Text)